hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/procedures/ShowTimelineProcedure.scala

```diff
@@ -0,0 +1,381 @@
+    val rollbackInfoMap = scala.collection.mutable.Map[String, scala.collection.mutable.ListBuffer[String]]()
+
+    val rollbackInstants = timeline.filter(instant =>
+      HoodieTimeline.ROLLBACK_ACTION.equalsIgnoreCase(instant.getAction)).getInstants
+
+    rollbackInstants.asScala.foreach { rollbackInstant =>
+      try {
+        if (rollbackInstant.isInflight) {
+          val instantToUse = metaClient.createNewInstant(
+            HoodieInstant.State.REQUESTED, rollbackInstant.getAction, rollbackInstant.requestedTime())
+          val metadata = timeline.readRollbackPlan(instantToUse)
+          val rolledBackInstant = metadata.getInstantToRollback.getCommitTime
+          rollbackInfoMap.getOrElseUpdate(rolledBackInstant, scala.collection.mutable.ListBuffer[String]())
+            .append(rollbackInstant.requestedTime())
+        } else {
+          val metadata = timeline.readRollbackMetadata(rollbackInstant)
+          metadata.getCommitsRollback.asScala.foreach { instant =>
+            rollbackInfoMap.getOrElseUpdate(instant, scala.collection.mutable.ListBuffer[String]())
+              .append(rollbackInstant.requestedTime())
+          }
+        }
+      } catch {
+        case _: Exception => // Skip invalid rollback instants
+      }
+    }
+    rollbackInfoMap.map { case (k, v) => k -> v.toList }.toMap
+  }
+
+  private def getRollbackInfo(instant: HoodieInstant, timeline: HoodieTimeline, rollbackInfoMap: Map[String, List[String]], metaClient: HoodieTableMetaClient): String = {
+    try {
+      if (HoodieTimeline.ROLLBACK_ACTION.equalsIgnoreCase(instant.getAction)) {
+        if (instant.isInflight) {
+          val instantToUse = metaClient.createNewInstant(
+            HoodieInstant.State.REQUESTED, instant.getAction, instant.requestedTime())
+          val metadata = timeline.readRollbackPlan(instantToUse)
+          s"Rolls back ${metadata.getInstantToRollback.getCommitTime}"
+        } else {
+          val metadata = timeline.readRollbackMetadata(instant)
+          val rolledBackInstants = metadata.getCommitsRollback.asScala.mkString(", ")
+          s"Rolled back: $rolledBackInstants"
+        }
+      } else {
+        val instantTimestamp = instant.requestedTime()
+        if (rollbackInfoMap.contains(instantTimestamp)) {
+          s"Rolled back by: ${rollbackInfoMap(instantTimestamp).mkString(", ")}"
+        } else {
+          null
+        }
+      }
+    } catch {
+      case _: Exception => null
+    }
+  }
+
+  private case class HoodieInstantWithModTime(state: HoodieInstant.State,
+                                              action: String,
+                                              requestedTime: String,
+                                              completionTime: String,
+                                              modificationTimeMs: Long
+                                             ) {
+    def getModificationTime: Long = modificationTimeMs
```

Comment:
Looks like we do not need this function `def getModificationTime: Long = modificationTimeMs`